Unify the core sync logic

tomneedham opened the pull request:
remove some code duplication and get the core to run a proper sync on the user metadata during a login

@DeepDiver1975 @PVince81 note that the account table introduced a change in behavior:
`setDisplayName()` was no longer called on the backend while `canChangeDisplayName()` would still check the backend if it implemented that function. With the account table we can always change display name, email, avatar ... but we may want to prevent that. `canChangeDisplayName()` already contains a check on the system value `allow_user_to_change_display_name`. `canChangeAvatar()` and `canChangePassword()` already exist, but email is missing.

Conceptually there are now two cases for setDisplayName:
1. End user triggered in the UI. Changeability may depend on the backend, eg LDAP is readonly.
2. Sync triggered.

The current User object code is intended only for the first case.

@tomneedham we need to add a codepath for the second case. AFAICT we should directly work on the accounts... the sync case bypasses any end user limitations.

DeepDiver1975 commented:
Tests hate you 😢

butonic commented on lib/private/User/Manager.php:
```diff
@@ -216,10 +221,13 @@ public function checkPassword($loginName, $password) {
 				if ($uid !== false) {
 					try {
 						$account = $this->accountMapper->getByUid($uid);
+						$this->syncService->setupAccount($account, $backend, $account->getUserId());
```
- [ ] move to a different place that is allso executed after login via apache / saml

tomneedham commented:
`CacheTest` seems to not find the displayname on a test user.. weird

tomneedham commented:
I most definitely just broke stuff - but I hope the logic is right.... LDAP login works, DB users work, saml seems to die.

tomneedham commented:
![screen shot 2017-11-29 at 20 54 36](https://user-images.githubusercontent.com/660805/33398636-8fec00d2-d547-11e7-8476-820a69a633de.png)
Safe to say stuff broke...

tomneedham commented:
Ok - fixed saml. But, for some reason the home path is not set so when you login with a new user they done have a home and you get some weird error about symlinks

tomneedham commented:
Fixed to use default home path if backend does not provide one. 

However - this _could_ break the home path during a sync if the backend suddently cannot provide a path or provides an invalid path.... I will add a check to not reset it

tomneedham commented:
- [ ] core should validate emails before setting them to oc_accounts.email

tomneedham commented:
- [ ] Might have killed some pre and post create user hooks

tomneedham commented:
Since the syncservice now does user creation, maybe the hooks wont be connected properly anymore...

tomneedham commented:
I did slightly re-order the hooks here, preCreateUser is now called after the validatePassword GenericEvent...

tomneedham commented on lib/private/User/Session.php:
```diff
@@ -554,18 +555,34 @@ private function loginWithToken($token) {
 		$uid = $backend->getCurrentUserId();
-		if ($uid === $this->getUser()->getUID()) {
+
+		if ($this->getUser() != null && $uid === $this->getUser()->getUID()) {
```
If session is active and current session uid matches the request user id then we are authenticated and have nothing more to do

tomneedham commented on lib/private/User/Session.php:
```diff
@@ -554,18 +555,34 @@ private function loginWithToken($token) {
+		// Die here if not valid
+		if(!$backend->isSessionActive()) {
+			return false;
+		}
```
This is redundant. OC_User::handleApacheAuth will only call this method if there is already an active IApacheBackend with an active session

tomneedham commented on lib/private/User/Session.php:
```diff
@@ -554,18 +555,34 @@ private function loginWithToken($token) {
+		try {
+			$syncService->createOrSyncAccount($uid, $backend);
+		} catch (\InvalidArgumentException $e) {
+			throw new LoginException('Uid collision2');
```
remove the 2

tomneedham commented on lib/private/User/Session.php:
```diff
@@ -554,18 +555,34 @@ private function loginWithToken($token) {
-	public function loginWithApache(\OCP\Authentication\IApacheBackend $backend) {
+	public function loginWithApache(IApacheBackend $backend) {
 
 		$uid = $backend->getCurrentUserId();
```
This can throw a HintException, should we catch and show as LoginException?

tomneedham commented on lib/private/User/SyncService.php:
```diff
@@ -178,15 +178,49 @@ public function syncAccount(Account $a, UserInterface $backend, $uid) {
+		// Check the backend is the same
+		if($account->getBackend() != get_class($backend)) {
+			// Then we have a uid collision
+			$this->logger->info("Uid collision when uid: $uid attempted to login");
```
log the exception

tomneedham commented on lib/private/User/SyncService.php:
```diff
@@ -178,15 +178,49 @@ public function syncAccount(Account $a, UserInterface $backend, $uid) {
+		if($account->getBackend() != get_class($backend)) {
+			// Then we have a uid collision
+			$this->logger->info("Uid collision when uid: $uid attempted to login");
+			throw new \InvalidArgumentException('Uid collision1');
```
remove the 1

tomneedham commented on lib/private/User/Manager.php:
```diff
@@ -353,9 +346,11 @@ public function createUser($uid, $password) {
 	 */
 	public function createUserFromBackend($uid, $password, $backend) {
 		$this->emit('\OC\User', 'preCreateUser', [$uid, '']);
```
password should be passed through here

tomneedham commented on lib/private/User/Manager.php:
```diff
@@ -353,9 +346,11 @@ public function createUser($uid, $password) {
 	 */
 	public function createUserFromBackend($uid, $password, $backend) {
```
deprecate interface, we dont want backends to be creating their own accounts whenever they fancy

tomneedham commented on lib/private/User/Manager.php:
```diff
@@ -353,9 +346,11 @@ public function createUser($uid, $password) {
+		try {
+			$account = $this->syncService->createOrSyncAccount($uid, get_class($backend));
+		} catch (\InvalidArgumentException $e) {
+			return null; // because thats what this method shoudl do apparently
```
add comment that this is logged already in the sync service

butonic commented on lib/private/User/Session.php:
```diff
@@ -557,9 +559,26 @@ private function loginWithToken($token) {
+		} else if (is_string($uidAndBackend)) {
+			$uid = $uidAndBackend;
+			if ($apacheBackend instanceof UserInterface) {
+				$backend = $apacheBackend;
```
,ore logging in case something goes wrong

tomneedham replied:
Done.

tomneedham commented on lib/base.php:
```diff
@@ -724,6 +725,23 @@ public static function init() {
+				self::$server->getAccountMapper()->update($account);
+			});
+		}
+	}
```
should now be able to remove this as sync is handled when each of the authentication types is attempted

tomneedham commented:
Still two test failures that look like they are to do with how the `UserTrait` confgiures a fake backend for the usermanager..

tomneedham commented:
Ah got it, the sync service isnt using the memory account mapper

tomneedham commented:
Added sync service to the server container. Added a method to override the mapper so unit tests can mock this or replace with the memory account mapper.

tomneedham commented:
Remaining problems are also related to no using the mock/dummy accountMapper - the sync service needs to know about this as well.

tomneedham commented:
rebased

tomneedham commented:
Dammit - this dav PR replaces IConfig in the server contain which breaks the sync service - why on earth does it replace the whole config.... https://github.com/owncloud/core/blob/ce4a435ed27db6f1ad4d123271898ea89be99744/apps/dav/tests/unit/Connector/Sabre/RequestTest/PartFileInRootUploadTest.php#L46

tomneedham commented:
Like some weird hash of unit and integration test, but not actually being either.

tomneedham commented:
Mmmm pushed a small hack for that test - @DeepDiver1975 is this ok?

tomneedham commented:
- [ ] TODO fix this line as well https://github.com/owncloud/core/blob/ce4a435ed27db6f1ad4d123271898ea89be99744/lib/private/legacy/api.php#L381

jvillafanez commented on lib/private/User/SyncService.php:
```diff
@@ -125,65 +130,188 @@ public function run(\Closure $callback) {
+	private function syncHome(Account $a, UserInterface $backend) {
+		// Home is handled differently, it should only be set on account creation, when there is no home already set
+		// Otherwise it could change on a sync and result in a new user folder being created
+		if($a->getHome() === null) {
```
I'm not sure if we should do something if the home is different between the account and the backend. As annoying as it will likely be, a warning message saying that we're refusing to update the home folder might be a good idea, specially if the admin expects the home to be updated. Since we don't have any command to change the home folder of a user, I think this is enough; the admin should revert the change in the backend.

tomneedham replied:
Makes sense to at least output something here - because the admin should be aware.

butonic commented on lib/private/User/Session.php:
```diff
@@ -551,6 +560,94 @@ private function loginWithToken($token) {
+			// completed before we can safely create the users folder.
+			// For example encryption needs to initialize the users keys first
+			// before we can create the user folder with the skeleton files
+			//OC_Hook::emit("OC_User", "post_login", ["uid" => $uid, 'password' => '']);
```
@tomneedham wrong hook? remove line?

tomneedham replied:
👍

tomneedham commented:
Started writing tests. Couldn't mock $backend->getHome because it doesnt exist in `UserInterface`. Talked to @DeepDiver1975 - so just went the whole way and added interfaces for providing a home and providing a displayname.

tomneedham commented:
So... I probably broke more stuff.

Commit:
Core controlled user metadata sync

mrow4a commented on lib/private/User/Session.php:
```diff
@@ -557,6 +566,91 @@ private function loginWithToken($token) {
+		}
+
+		// Now we try to create the account or sync
+		$this->userSyncService->createOrSyncAccount($uid, $backend);
```
This means each new login with result in sync, right?

mrow4a commented on lib/private/User/SyncService.php:
```diff
@@ -125,65 +132,197 @@ public function run(\Closure $callback) {
+	 * @param UserInterface $backend of the user
+	 * @return Account
+	 */
+	public function syncAccount(Account $a, UserInterface $backend) {
```
Isnt it 10 calls to LDAP just for one login? @DeepDiver1975 @butonic @tomneedham

mrow4a replied:
I guess no way it can be 1 call right?

mrow4a commented:
Well have to admin, this PR cleaned up a lot of things... code looks now... like a code ;d

phil-davis commented on lib/private/User/SyncService.php:
```diff
@@ -89,32 +93,35 @@ public function getNoLongerExistingUsers(\Closure $callback) {
+					$this->syncAccount($a, $backend);
 					$this->mapper->update($a);
 				} catch(DoesNotExistException $ex) {
-					$a = $this->createNewAccount($uid);
```
change in this line fixed/reverted by https://github.com/owncloud/core/pull/30367
